assets/src/edit-story/app/media/media3p/useFetchMediaEffect.js

```diff
@@ -53,41 +57,66 @@ export default function useFetchMediaEffect({
     actions: { listMedia, listCategoryMedia },
   } = useMedia3pApi();
 
+  const previousProps = usePrevious({
+    pageToken,
+    searchTerm,
+    selectedCategoryId,
+  });
+
   const { showSnackbar } = useSnackbar();
 
   useEffect(() => {
     async function fetch() {
       fetchMediaStart({ provider, pageToken });
       try {
-        let media, nextPageToken;
+        let media, newNextPageToken;
         if (selectedCategoryId) {
-          ({ media, nextPageToken } = await listCategoryMedia({
-            provider,
-            selectedCategoryId,
-            pageToken,
-          }));
+          ({ media, nextPageToken: newNextPageToken } = await listCategoryMedia(
+            {
+              provider,
+              selectedCategoryId,
+              pageToken,
+            }
+          ));
         } else {
-          ({ media, nextPageToken } = await listMedia({
+          ({ media, nextPageToken: newNextPageToken } = await listMedia({
             provider,
             searchTerm,
             pageToken,
           }));
         }
-        fetchMediaSuccess({ provider, media, pageToken, nextPageToken });
+        fetchMediaSuccess({
+          provider,
+          media,
+          pageToken,
+          nextPageToken: newNextPageToken,
+        });
       } catch {
         fetchMediaError({ provider, pageToken });
         showSnackbar({ message: getFetchMediaErrorMessage(provider) });
       }
     }
 
-    if (provider === selectedProvider) {
+    // If we switched provider tab, and we already had loaded a page there, we
+    // don't load media again.
+    const somethingChanged =
+      previousProps &&
+      (pageToken != previousProps.pageToken ||
+        searchTerm != previousProps.searchTerm ||
+        selectedCategoryId != previousProps.selectedCategoryId);
+    const firstFetchOrSomethingChanged =
+      (!pageToken && !nextPageToken) || somethingChanged;
```

Comment:
I don't think you need 'nextPageToken' to know if it's the first fetch. pageToken will be undefined if the first page is being fetched.

Reply:
I think if you've already fetched data, nextPageToken is not null but pageToken is. We don't want to fetch if we've already fetched in that case, so I need to check both?

Reply:
ahh, you are right.
I wonder if that's a problem, ie. if there is <= 1 page of images, then pageToken and nextPageToken will both be undefined, even if the first fetch has already completed.

Reply:
Good point, I've modified the logic to depend on isMediaLoading and isMediaLoaded instead. If both are false, that means that no media has been fetched yet.